Subject: Work around segfault on ppc64el
Author: Christoph Biedl <debian.axhn@manchmal.in-ulm.de>
Date: 2021-12-20
Bug-Debian: https://bugs.debian.org/995393
Forwarded: Yes

    For whatever reason the generated code segfaults on ppc64el when
    built with the usual optimizations. The root cause is not clear,
    might be a compiler bug or the result of improperly generated
    function prototypes.

    As a workaround, disable optimizations for the affected function.

    This should be re-visted upon any major compiler release whether
    it's still needed.

--- a/libfakeroot.c
+++ b/libfakeroot.c
@@ -2596,7 +2596,11 @@
 #endif
 
 #ifdef HAVE_OPENAT
-int openat(int dir_fd, const char *pathname, int flags, ...)
+int
+#if defined(__powerpc__) && defined(__powerpc64__) && __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+__attribute__((optimize("O0")))
+#endif
+openat(int dir_fd, const char *pathname, int flags, ...)
 {
 	mode_t mode;
 
